File: 2018/d5.py
from utils import readfile

inp = readfile("d5_input.txt").strip()
inp = list(inp)

i = 0
while True:
    if i == len(inp)-2:
        # end of list
        break
    elif abs(ord(inp[i]) - ord(inp[i+1])) == 32:
        # same polymer type (same char) of diff polarity (upper vs lowercase) are next to each other

        # use slice to delete multiple contiguous elements from list
        del inp[i:i+2]
        # since we delete the element at i this wont delete the original i+1 but i+2
        # del inp[i]
        # del inp[i+1]
        if i > 0:
            i += -1
    else:
        i += 1
print("Part1:", len(inp))

lengths = []
for c in list("abcdefghijklmnopqrstuvwxyz"):
    # replace one char/polymer type in both cases/polarity
    inp_repl = "".join(inp).replace(c, "").replace(c.upper(), "")
    inp_repl = list(inp_repl)
    i = 0
    while True:
        if i == len(inp_repl)-2:
            break
        elif abs(ord(inp_repl[i]) - ord(inp_repl[i+1])) == 32:
            # use slice to delete multiple contiguous elements from list
            del inp_repl[i:i+2]
            if i > 0:
                i += -1
        else:
            i += 1
    lengths.append((c, len(inp_repl)))
print("Part2:", min(lengths, key=lambda l: l[1]))

File: 2018/d5.go
package main

import (
    "fmt"
    "bufio"
    "os"
    "strings"
)

func main() {
    b, err := os.Open("d5_input.txt")
    if err != nil {
        fmt.Println(err)
    }
    scanner := bufio.NewScanner(b)
    // Scan advances the Scanner to the next token, which will then be
    // available through the Bytes or Text method
    scanner.Scan()
    // slice of byte
    inp := []byte(strings.TrimSpace(scanner.Text()))
    if err := scanner.Err(); err != nil {
        fmt.Println("Error reading from file input:", err)
    }
    // go passes almost? everything by value so an assignment is like a copy operation
    // but it doesnt work for slices (would work for arrays or structs) and inp is a slice of bytes
    // inp_copy = inp
    // The copy function supports copying between slices of different lengths (!!! it will copy only up to the smaller number of elements !!!)
    // so make them the same size by passing len(inp)
    inp_copy := make([]byte, len(inp))
    // dest, src
    copy(inp_copy, inp)
    i := 0
    for {
        if i > len(inp_copy)-2 {
            break
        } else if int(inp_copy[i]) - int(inp_copy[i+1]) == 32 || int(inp_copy[i]) - int(inp_copy[i+1]) == -32 {
            // del i and i+1 from array by appending all elements beginning with i+2
            // to slice ending before i
            inp_copy = append(inp_copy[:i], inp_copy[i+2:]...)
            if i > 0 {
                i--
            }
        } else {
            i++
        }
    }
    fmt.Println("Part1:", len(inp_copy))

    var lengths [26][2]int
    for j := range lengths {
        c := 97 + j
        inp_str := strings.Replace(strings.Replace(string(inp), string(c), "", -1), string(c-32), "", -1)
        inp_copy = []byte(inp_str)
        i := 0
        for {
            if i > len(inp_copy)-2 {
                break
            } else if (int(inp_copy[i]) - int(inp_copy[i+1]) == 32) || (int(inp_copy[i]) - int(inp_copy[i+1]) == -32) {
                // fmt.Println(string(inp_copy[i-1]), string(inp_copy[i]), string(inp_copy[i+1]), string(inp_copy[i+2]))
                // del i and i+1 from array by appending all elements beginning with i+2
                // to slice ending before i; ... unpacks slice into var args
                inp_copy = append(inp_copy[:i], inp_copy[i+2:]...)
                if i > 0 {
                    i--
                }
            } else {
                i++
            }
        }        
        lengths[j][0] = c
        lengths[j][1] = len(inp_copy)
    }
    min := -1
    for _, c_len := range lengths {
        if c_len[1] < min {
            min = c_len[1]
        } else if min == -1 {
            min = c_len[1]
        }
    }
    fmt.Println("Part2:", min)
    }
